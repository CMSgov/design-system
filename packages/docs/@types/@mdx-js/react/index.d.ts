declare module '@mdx-js/react' {
  const MDXProvider: React.ComponentType;
  export { MDXProvider };
}
