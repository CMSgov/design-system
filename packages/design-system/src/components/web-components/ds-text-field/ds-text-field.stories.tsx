import WebComponentDocTemplate from '../../../../../../.storybook/docs/WebComponentDocTemplate.mdx';
import { action } from '@storybook/addon-actions';
import { useEffect } from 'react';
import { webComponentDecorator } from '../storybook';
import './ds-text-field';

export default {
  title: 'Web Components/ds-text-field',
  argTypes: {
    'class-name': {
      description: 'Additional classes to be added to the root element.',
      control: 'text',
    },
    disabled: { description: 'Disables the entire field.', control: 'boolean' },
    'error-message': {
      description: 'Enable the error state by providing an error message',
      control: 'text',
    },
    'error-id': {
      description:
        'The ID of the error message applied to this field. If none is provided, the id will be derived from the `root-id` attribute.',
      control: 'text',
    },
    'error-placement': {
      description: 'Location of the error message relative to the field input',
      options: [undefined, 'top', 'bottom'],
      control: { type: 'radio' },
    },
    'field-class-name': {
      description: 'Additional classes to be added to the input element',
      control: 'text',
    },
    hint: {
      description: 'Hint text or HTML',
      control: 'text',
    },
    'hint-class-name': {
      control: 'text',
      description: 'Custom CSS class name(s) for the hint element',
    },
    'hint-id': {
      description: 'The ID of the hint element',
      control: 'text',
    },
    inversed: {
      description: 'Set to "true" to apply the "inverse" theme',
      control: 'boolean',
    },
    label: {
      description: 'Label text or HTML.',
      control: 'text',
    },
    'label-class-name': {
      description: 'Additional classes to be added to the field label',
      control: 'text',
    },
    'label-id': {
      description:
        "A unique `id` to be used on the field label. If one isn't provided, a unique ID will be generated.",
      control: 'text',
    },
    'label-mask': {
      description:
        'Providing a mask type here will turn the text field into a label-masked field, where the user input is formatted in a label as the user types and then the input field itself is automatically formatted when the user shifts focus away from the input. See [Label-masked field](https://design.cms.gov/components/text-field/label-masked-field/) documentation page for more information.',
      options: [undefined, 'phone', 'zip', 'ssn', 'currency'],
      control: { type: 'radio' },
    },
    name: {
      description: "The `input` field's `name` attribute.",
      control: 'text',
    },
    'requirement-label': {
      control: 'text',
      description:
        'Text showing the requirement (ie. "Optional", or "Required").\nIn most cases, this should be used to indicate which fields are optional.\nSee the [form guidelines](https://design.cms.gov/patterns/Forms/forms/) for more info.',
    },
    size: {
      description: 'Sets the max-width of the input either to `"small"` or `"medium"`',
      options: [undefined, 'medium', 'small'],
      control: { type: 'radio' },
    },
    value: {
      description: "The `input` field's `value` attribute",
      control: 'text',
    },
    'root-id': {
      control: 'text',
      description:
        "A unique ID for this element. A unique ID will be generated if one isn't provided.",
    },
  },
  args: {
    label: 'Enter some text.',
    name: 'text-field',
  },
  parameters: {
    docs: {
      page: WebComponentDocTemplate,
      description: {
        component: `For information about how and when to use this component, [refer to its full documentation page](https://design.cms.gov/components/text-field/).`,
      },
    },
  },
  decorators: [webComponentDecorator],
};

const Template = (args) => {
  useEffect(() => {
    const onChange = (event) => {
      action('ds-change')(event);
    };
    const onBlur = (event) => {
      action('ds-blur')(event);
    };
    const textField = document.querySelector('ds-text-field');
    textField.addEventListener('ds-change', onChange);
    textField.addEventListener('ds-blur', onBlur);
    return () => {
      textField.removeEventListener('ds-change', onChange);
      textField.removeEventListener('ds-blur', onBlur);
    };
  });

  return <ds-text-field {...args} />;
};

export const Default = Template.bind({});

export const MultilineField = {
  render: Template,
  args: {
    multiline: 'true',
    rows: '3',
  },
};
