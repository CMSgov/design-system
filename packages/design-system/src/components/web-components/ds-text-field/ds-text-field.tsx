import { define } from '../preactement/define';
import { TextField, TextFieldProps } from '../../TextField';
import { parseBooleanAttr } from '../wrapperUtils';
import { formAttrs } from '../shared-attributes/form';
import { PHONE_MASK, SSN_MASK, ZIP_MASK, CURRENCY_MASK } from '../../TextField/useLabelMask';

const attributes = [
  // Using the lowercase HTML attribute name rather than `auto-focus` so it's
  // more natural. There's no reason for us to worry about name collisions.
  'autofocus',
  'aria-disabled',
  'class-name',
  'disabled',
  'field-class-name',
  'inversed',
  'label-mask',
  'multiline',
  'name',
  'numeric',
  'value',
  'root-id',
  'rows',
  'size',
  'type',
  ...formAttrs,
];

function getMaskFunction(value: string) {
  switch (value) {
    case 'phone':
      return PHONE_MASK;
    case 'zip':
      return ZIP_MASK;
    case 'ssn':
      return SSN_MASK;
    case 'currency':
      return CURRENCY_MASK;
  }
  return undefined;
}

interface WrapperProps
  extends Omit<
    TextFieldProps,
    'autoFocus' | 'ariaDisabled' | 'disabled' | 'inversed' | 'labelMask' | 'multiline'
  > {
  autofocus?: string;
  ariaDisabled?: string;
  disabled?: string;
  inversed?: string;
  labelMask?: string;
  multiline?: string;
  rootId?: string;
}

const Wrapper = ({ autofocus, rootId, ...otherProps }: WrapperProps) => (
  <TextField
    {...otherProps}
    autoFocus={parseBooleanAttr(autofocus)}
    aria-disabled={parseBooleanAttr(otherProps.ariaDisabled)}
    disabled={parseBooleanAttr(otherProps.disabled)}
    id={rootId}
    inversed={parseBooleanAttr(otherProps.inversed)}
    labelMask={getMaskFunction(otherProps.labelMask)}
    multiline={parseBooleanAttr(otherProps.multiline)}
  />
);

define('ds-text-field', () => Wrapper, { attributes, events: ['onChange', 'onBlur'] });
