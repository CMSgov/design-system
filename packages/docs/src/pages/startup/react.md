---
title: Using React components
weight: 3
---

The examples below assume you've installed the design system using NPM and have already setup your build system.

<h4 class="ds-h4 ds-u-font-size--base">Default imports</h4>

Individual components can be imported from their individual export file.

```jsx
import Button from '@cmsgov/design-system-core/dist/components/Button/Button';
```

<h4 class="ds-h4 ds-u-font-size--base">Named imports</h4>

Components can also be imported using the shorter syntax below.

<div class="ds-c-alert ds-c-alert--warn ds-u-margin-bottom--2 ds-u-font-size--small">
  <div class="ds-c-alert__body">
    <h3 class="ds-c-alert__heading ds-u-font-size--base">Performance note</h3>
    <p class="ds-c-alert__text">
      This approach may result in a much larger file than you intend. Depending on what module bundler you use, all of the design system's React components may be included in the bundled file even if you didn't specifically import them. This can be avoided by enabling features like [tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/).
    </p>
  </div>
</div>

```jsx
import { Button } from '@cmsgov/design-system-core';
```

[View an example](https://github.com/CMSgov/design-system/blob/master/examples/react-app/src/scripts/index.js)

<h2 id="examples" class="ds-h2 ds-u-color--primary-darker">Examples</h2>

Additional examples of the design system in use can be viewed on GitHub. These projects demonstrate the various ways you can incorporate the design system into your development process and various use cases.

<a href="https://github.com/CMSgov/design-system/tree/master/examples/" class="ds-c-button">Browse example projects</a>

<h2 id="need-help" class="ds-h2 ds-u-color--primary-darker">Need help or ran into an issue?</h2>

If you're having trouble installing or setting up the design system, or if you think you've found a bug, feel free to [open an issue on GitHub](https://github.com/CMSgov/design-system/issues).
