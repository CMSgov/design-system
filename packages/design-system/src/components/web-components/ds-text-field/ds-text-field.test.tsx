import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import './ds-text-field';

/* eslint-disable @typescript-eslint/no-namespace */
declare global {
  namespace JSX {
    interface IntrinsicElements {
      'ds-text-field': any;
    }
  }
}
/* eslint-enable */

function renderTextField(props = {}) {
  return render(<ds-text-field {...props} />);
}

describe('ds-text-field', function () {
  it('renders text-field', () => {
    const { asFragment } = renderTextField({ id: 'static-id' });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders disabled input', () => {
    renderTextField({ disabled: 'true' });
    const input = screen.getByRole('textbox');
    expect(input).toHaveAttribute('disabled');
  });

  it('does not render disabled when passed false', () => {
    renderTextField({ disabled: 'false' });
    const input = screen.getByRole('textbox');
    expect(input).not.toHaveAttribute('disabled');
  });

  it('applies additional classes to the wrapper', () => {
    const { container } = renderTextField({ 'class-name': 'foobar' });
    const wrapper = container.querySelector('.ds-u-clearfix');
    expect(wrapper).toHaveClass('foobar');
  });

  it('applies additional classes to the input', () => {
    renderTextField({ 'field-class-name': 'foobar' });
    const input = screen.getByRole('textbox');
    expect(input).toHaveClass('foobar');
  });

  it('applies requirement-label to hint text', () => {
    const { container } = renderTextField({ 'requirement-label': 'Optional' });
    const input = screen.getByRole('textbox');
    const hintId = input.getAttribute('aria-describedby');
    const hint = container.querySelector(`#${hintId}`);
    expect(hint).toContainHTML('Optional');
  });

  it('renders an error message', () => {
    const { container } = renderTextField({ 'error-message': 'Uh-oh!' });
    const input = screen.getByRole('textbox');
    const errorId = input.getAttribute('aria-describedby');
    const error = container.querySelector(`#${errorId}`);
    expect(error).toContainHTML('Uh-oh!');
  });

  it('applies inverse class', () => {
    const { container } = renderTextField({
      inversed: 'true',
      hint: 'Hello',
      'error-message': 'Ahh!!',
    });
    const inversedLabel = container.querySelector('.ds-c-label--inverse');
    const inversedHint = container.querySelector('.ds-c-hint--inverse');
    const inversedError = container.querySelector('.ds-c-inline-error--inverse');
    expect(inversedLabel).toBeInTheDocument();
    expect(inversedHint).toBeInTheDocument();
    expect(inversedError).toBeInTheDocument();
  });

  it('applies size classes', () => {
    renderTextField({ size: 'small' });
    const input = screen.getByRole('textbox');
    expect(input.classList.contains('ds-c-field--small')).toBe(true);
  });

  it('supports multiline text fields', () => {
    renderTextField({ multiline: 'true', rows: '3' });
    const textArea = screen.getByRole('textbox');
    expect(textArea.tagName).toEqual('TEXTAREA');
    expect(textArea).toHaveAttribute('rows', '3');
  });

  it('fires a custom ds-change event', () => {
    renderTextField();

    const textFieldRoot = document.querySelector('ds-text-field');
    const mockHandler = jest.fn();
    textFieldRoot.addEventListener('ds-change', mockHandler);

    const input = screen.getByRole('textbox');
    userEvent.click(input);
    userEvent.keyboard('a');

    expect(mockHandler).toHaveBeenCalledTimes(1);
    textFieldRoot.removeEventListener('ds-change', mockHandler);
  });

  it('fires a custom ds-blur event', async () => {
    renderTextField();

    const textFieldRoot = document.querySelector('ds-text-field');
    const onBlur = jest.fn();
    const onChange = jest.fn();
    textFieldRoot.addEventListener('ds-blur', onBlur);
    textFieldRoot.addEventListener('ds-change', onChange);

    const input = screen.getByRole('textbox');
    userEvent.click(input);
    userEvent.tab();

    const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
    await act(async () => {
      await sleep(40);
    });

    expect(onBlur).toHaveBeenCalledTimes(1);
    expect(onChange).not.toHaveBeenCalled();
    textFieldRoot.removeEventListener('ds-blur', onBlur);
    textFieldRoot.removeEventListener('ds-change', onChange);
  });

  it('formats a phone number on blur', () => {
    renderTextField({ 'label-mask': 'phone' });

    const input = screen.getByRole('textbox') as HTMLInputElement;
    userEvent.click(input);
    userEvent.type(input, '1234567890');
    userEvent.tab();

    expect(input.value).toEqual('123-456-7890');
  });
});
