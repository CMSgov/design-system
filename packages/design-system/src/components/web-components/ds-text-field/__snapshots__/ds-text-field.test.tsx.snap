// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ds-text-field renders text-field 1`] = `
<DocumentFragment>
  <ds-text-field
    id="static-id"
  >
    <div
      class="ds-u-clearfix"
    >
      <label
        class="ds-c-label"
        for="text-field--P0-0"
        id="text-field--P0-0__label"
      />
      <p
        aria-atomic="true"
        aria-live="assertive"
        class="ds-c-inline-error"
        id="text-field--P0-0__error"
      />
      <input
        aria-disabled="false"
        aria-invalid="false"
        class="ds-c-field"
        id="text-field--P0-0"
        type="text"
      />
    </div>
    <template />
  </ds-text-field>
</DocumentFragment>
`;
